Butterfly.SqlServer/SqlServerDatabase.cs

```diff
@@ -0,0 +1,347 @@
+﻿/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+using Butterfly.Core.Database;
+using NLog;
+using System;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Data.SqlClient;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+using Dict = System.Collections.Generic.Dictionary<string, object>;
+
+namespace Butterfly.SqlServer
+{
+	/// <inheritdoc/>
+	public class SqlServerDatabase : BaseDatabase
+	{
+		private static readonly Logger logger = LogManager.GetCurrentClassLogger();
+
+		private static Regex LIMIT_REGEX = new Regex(@"^(SELECT).+(LIMIT\s+(.+?))$", RegexOptions.IgnoreCase | RegexOptions.Singleline);
+
+		public SqlServerDatabase(string connectionString) : base(connectionString) {
+		}
+		
+		public override bool CanJoin => true;
+
+		public override bool CanFieldAlias => true;
+
+		protected override BaseTransaction CreateTransaction()
+		{
+			return new SqlServerTransaction(this);
+		}
+
+		protected override async Task<Dict[]> DoQueryRowsAsync(string storedProcedureName, Dict executableParams)
+		{
+			var result = await ExecuteCommandAsync<Dict[]>(async c =>
+			{
+				var rows = new List<Dict>();
+				using (var reader = await c.ExecuteReaderAsync())
+				{
+					while (await reader.ReadAsync())
+					{
+						var row = new Dict();
+						for (int i = 0; i < reader.FieldCount; i++)
+						{
+							row[reader.GetName(i)] = reader[i];
+						}
+					}
+				}
+
+				return rows.ToArray();
+			}, storedProcedureName, executableParams);
+
+			return result;
+		}
+
+		protected override async Task<Dict[]> DoSelectRowsAsync(string executableSql, Dict executableParams)
+		{
+			var sql = ConvertLimit(executableSql);
+
+			var result = await ExecuteCommandAsync<Dict[]>(async c =>
+			{
+				var rows = new List<Dict>();
+				using (var reader = await c.ExecuteReaderAsync())
+				{
+					var columns = reader.GetColumnSchema();
+					while (await reader.ReadAsync())
+					{
+						var row = new Dict();
+						foreach (var column in columns)
+						{
+							row[column.ColumnName] = reader[column.ColumnName];
+						}
+
+						rows.Add(row);
+					}
+				}
+
+				return rows.ToArray();
+			}, sql, executableParams);
+
+			return result;
+		}
+
+		private string ConvertLimit(string sql)
```

Comment:
The need for this function, as well as the extra `executableSql += " SELECT SCOPE_IDENTITY();";` in `SqlServerTransaction.DoInsertAsync()`, should probably be removed once/if we override the `SelectStatement` and `InsertStatement`